### PR TITLE
VB-2470 add validation around updates of valid to date to allow updates if visits are cancelled.

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
## What does this pull request do?

Updates validation around valid to date to looked for booked or reserved visits.
## What is the intent behind these changes?

Admin API improvements.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt [hunk 1]
```diff
@@ -79,11 +79,13 @@ interface VisitRepository : JpaRepository<Visit, Long>, JpaSpecificationExecutor
   @Query(
     "SELECT count(v) > 0 FROM Visit v " +
       "WHERE v.sessionTemplateReference = :sessionTemplateReference AND " +
-      "(cast(:visitsFromDate as date)  is null or cast(v.visitStart as date) >= :visitsFromDate) ",
+      "(cast(:visitsFromDate as date)  is null or cast(v.visitStart as date) >= :visitsFromDate) AND " +
+      "((:#{#visitStatuses == null} = true) OR v.visitStatus IN (:visitStatuses))",
   )
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:84]: new syntax for me # does this check for an empty list?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/UpdateSessionTemplateValidator.kt [hunk 2]
```diff
@@ -64,9 +65,10 @@ class UpdateSessionTemplateValidator(
       if (newValidToDate != existingValidToDate) {
         // if the new validToDate is not null or before existing validToDate
         if ((newValidToDate != null && existingValidToDate == null) || (newValidToDate != null && newValidToDate.isBefore(existingValidToDate))) {
-          // check if there are any visits (any visit status) after the new valid to date
-          if (visitRepository.hasVisitsForSessionTemplate(existingSessionTemplate.reference, newValidToDate.plusDays(1))) {
-            return "Cannot update session valid to date to $newValidToDate for session template - ${existingSessionTemplate.reference} as there are visits associated with this session template after $newValidToDate."
+          // check if there are any booked or reserved visits (any visit status) after the new valid to date
+          val visitStatues = listOf<VisitStatus>(VisitStatus.BOOKED, VisitStatus.RESERVED, VisitStatus.CHANGING)
+          if (visitRepository.hasVisitsForSessionTemplate(existingSessionTemplate.reference, newValidToDate.plusDays(1), visitStatues)) {
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/UpdateSessionTemplateValidator.kt:70]: would make this a constant or put it directly in the query, could also put it in it's own local method hasVisitsForSessionTemplate

**Review by AledWEvans — COMMENTED** [review]
Some comments, try to encapsulate the the functionality into functions

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt (hunk 1 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:84]: Yes, got it from here - https://hibernate.atlassian.net/browse/HHH-15968 but maybe as you suggested adding them to the query might make more sense.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/UpdateSessionTemplateValidator.kt (hunk 2 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/UpdateSessionTemplateValidator.kt:70]: should I make 2 methods? for some instances we might need Cancel in some cases e.g. if they are trying to change the from date but any type of visits exists

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt (hunk 1 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:84]: >add them to the query itself
I think thats the general approach we have used in VSiP.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/UpdateSessionTemplateValidator.kt (hunk 2 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/UpdateSessionTemplateValidator.kt:70]: If we need 2 now, sure!

**Review by AledWEvans — APPROVED** [review]
LGTM

**Pull request merged**